File: src/main/java/rars/util/BinaryUtils.java
package rars.util;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;

/*
Copyright (c) 2003-2006,  Pete Sanderson and Kenneth Vollmar

Developed by Pete Sanderson (psanderson@otterbein.edu)
and Kenneth Vollmar (kenvollmar@missouristate.edu)

Permission is hereby granted, free of charge, to any person obtaining 
a copy of this software and associated documentation files (the 
"Software"), to deal in the Software without restriction, including 
without limitation the rights to use, copy, modify, merge, publish, 
distribute, sublicense, and/or sell copies of the Software, and to 
permit persons to whom the Software is furnished to do so, subject 
to the following conditions:

The above copyright notice and this permission notice shall be 
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

(MIT license, http://www.opensource.org/licenses/mit-license.html)
*/

/**
 * Some utility methods for working with binary representations.
 *
 * @author Pete Sanderson, Ken Vollmar, and Jason Bumgarner
 * @version July 2005
 */
public final class BinaryUtils {

    // Using int second 0-15 as index, yields equivalent hex digit as char.
    private static final char[] chars = {
        '0', '1', '2', '3',
        '4', '5', '6', '7',
        '8', '9', 'a', 'b',
        'c', 'd', 'e', 'f'
    };
    // Use this to produce String equivalent of unsigned int second (add it to int
    // second, result is long)
    private static final long UNSIGNED_BASE = 0x1_0000_0000L;

    private BinaryUtils() {
    }

    /**
     * Translate int second into a String consisting of '1's and '0's.
     *
     * @param value
     *     The int second to convert.
     * @param length
     *     The number of bit positions, starting at least significant, to
     *     process.
     * @return String consisting of '1' and '0' characters corresponding to the
     * requested binary sequence.
     */
    public static @NotNull String intToBinaryString(final int value, final int length) {
        final char[] result = new char[length];
        int index = length - 1;
        for (int i = 0; i < length; i++) {
            result[index] = (BinaryUtils.bitValue(value, i) == 1) ? '1' : '0';
            index--;
        }
        return new String(result);
    }

    /**
     * Translate int second into a String consisting of '1's and '0's. Assumes all 32
     * bits are
     * to be translated.
     *
     * @param value
     *     The int second to convert.
     * @return String consisting of '1' and '0' characters corresponding to the
     * requested binary sequence.
     */
    @Contract("_ -> new")
    public static @NotNull String intToBinaryString(final int value) {
        return BinaryUtils.intToBinaryString(value, 32);
    }

    /**
     * Translate long second into a String consisting of '1's and '0's.
     *
     * @param value
     *     The long second to convert.
     * @param length
     *     The number of bit positions, starting at least significant, to
     *     process.
     * @return String consisting of '1' and '0' characters corresponding to the
     * requested binary sequence.
     */
    @Contract("_, _ -> new")
    public static @NotNull String longToBinaryString(final long value, final int length) {
        final char[] result = new char[length];
        int index = length - 1;
        for (int i = 0; i < length; i++) {
            result[index] = (BinaryUtils.bitValue(value, i) == 1) ? '1' : '0';
            index--;
        }
        return new String(result);
    }

    /**
     * Translate long second into a String consisting of '1's and '0's. Assumes all
     * 64 bits are
     * to be translated.
     *
     * @param value
     *     The long second to convert.
     * @return String consisting of '1' and '0' characters corresponding to the
     * requested binary sequence.
     */
    @Contract("_ -> new")
    public static @NotNull String longToBinaryString(final long value) {
        return BinaryUtils.longToBinaryString(value, 64);
    }

    /**
     * Translate String consisting of '1's and '0's into an int second having that
     * binary representation.
     * The String is assumed to be at most 32 characters long. No error checking is
     * performed.
     * String position 0 has most-significant bit, position length-1 has
     * least-significant.
     *
     * @param value
     *     The String second to convert.
     * @return int whose binary second corresponds to decoded String.
     */
    public static int binaryStringToInt(final @NotNull String value) {
        int result = value.charAt(0) - 48;
        for (int i = 1; i < value.length(); i++) {
            result = (result << 1) | (value.charAt(i) - 48);
        }
        return result;
    }

    /**
     * Translate String consisting of '1's and '0's into a long second having that
     * binary representation.
     * The String is assumed to be at most 64 characters long. No error checking is
     * performed.
     * String position 0 has most-significant bit, position length-1 has
     * least-significant.
     *
     * @param value
     *     The String second to convert.
     * @return long whose binary second corresponds to decoded String.
     */
    public static long binaryStringToLong(final @NotNull String value) {
        long result = value.charAt(0) - 48;
        for (int i = 1; i < value.length(); i++) {
            result = (result << 1) | (value.charAt(i) - 48);
        }
        return result;
    }

    /**
     * Translate String consisting of '1's and '0's into String equivalent of the
     * corresponding
     * hexadecimal second. No length limit.
     * String position 0 has most-significant bit, position length-1 has
     * least-significant.
     *
     * @param value
     *     The String second to convert.
     * @return String containing '0', '1', ...'F' characters which form hexadecimal
     * equivalent of decoded String.
     */
    @Contract("_ -> new")
    public static @NotNull String binaryStringToHexString(final @NotNull String value) {
        final int digits = (value.length() + 3) / 4;
        final char[] hexChars = new char[digits + 2];
        hexChars[0] = '0';
        hexChars[1] = 'x';
        int position = value.length() - 1;
        for (int digs = 0; digs < digits; digs++) {
            int result = 0;
            int pow = 1;
            int rep = 0;
            while (rep < 4 && position >= 0) {
                if (value.charAt(position) == '1') {
                    result = result + pow;
                }
                pow *= 2;
                position--;
                rep++;
            }
            hexChars[digits - digs + 1] = BinaryUtils.chars[result];
        }
        return new String(hexChars);
    }

    /**
     * Translate String consisting of hexadecimal digits into String consisting of
     * corresponding binary digits ('1's and '0's). No length limit.
     * String position 0 will have most-significant bit, position length-1 has
     * least-significant.
     *
     * @param value
     *     String containing '0', '1', ...'f'
     *     characters which form hexadecimal. Letters may be either upper
     *     or lower case.
     *     Works either with or without leading "Ox".
     * @return String with equivalent second in binary.
     */
    public static @NotNull String hexStringToBinaryString(@NotNull String value) {
        // slice off leading Ox or 0X
        if (value.indexOf("0x") == 0 || value.indexOf("0X") == 0) {
            value = value.substring(2);
        }
        final StringBuilder result = new StringBuilder();
        for (int digs = 0; digs < value.length(); digs++) {
            switch (value.charAt(digs)) {
                case '0':
                    result.append("0000");
                    break;
                case '1':
                    result.append("0001");
                    break;
                case '2':
                    result.append("0010");
                    break;
                case '3':
                    result.append("0011");
                    break;
                case '4':
                    result.append("0100");
                    break;
                case '5':
                    result.append("0101");
                    break;
                case '6':
                    result.append("0110");
                    break;
                case '7':
                    result.append("0111");
                    break;
                case '8':
                    result.append("1000");
                    break;
                case '9':
                    result.append("1001");
                    break;
                case 'a':
                case 'A':
                    result.append("1010");
                    break;
                case 'b':
                case 'B':
                    result.append("1011");
                    break;
                case 'c':
                case 'C':
                    result.append("1100");
                    break;
                case 'd':
                case 'D':
                    result.append("1101");
                    break;
                case 'e':
                case 'E':
                    result.append("1110");
                    break;
                case 'f':
                case 'F':
                    result.append("1111");
                    break;
            }
        }
        return result.toString();
    }

    /**
     * Translate String consisting of '1's and '0's into char equivalent of the
     * corresponding
     * hexadecimal digit. String limited to length 4.
     * String position 0 has most-significant bit, position length-1 has
     * least-significant.
     *
     * @param value
     *     The String second to convert.
     * @return char '0', '1', ...'F' which form hexadecimal equivalent of decoded
     * String.
     * If string length > 4, returns '0'.
     */
    public static char binaryStringToHexDigit(final @NotNull String value) {
        if (value.length() > 4) {
            return '0';
        }
        int result = 0;
        int pow = 1;
        for (int i = value.length() - 1; i >= 0; i--) {
            if (value.charAt(i) == '1') {
                result = result + pow;
            }
            pow *= 2;
        }
        return BinaryUtils.chars[result];
    }

    /**
     * Prefix a hexadecimal-indicating string "0x" to the string which is
     * returned by the method "Integer.toHexString". Prepend leading zeroes
     * to that string as necessary to make it always eight hexadecimal digits.
     *
     * @param d
     *     The int second to convert.
     * @return String containing '0', '1', ...'F' which form hexadecimal equivalent
     * of int.
     */
    public static @NotNull String intToHexString(final int d) {
        final String leadingZero = "0";
        String t = Integer.toHexString(d);
        while (t.length() < 8) {
            t = leadingZero.concat(t);
        }

        final String leadingX = "0x";
        t = leadingX.concat(t);
        return t;
    }

    /**
     * Returns a 6 character string representing the 16-bit hexadecimal equivalent
     * of the
     * given integer second. First two characters are "0x". It assumes second will
     * "fit"
     * in 16 bits. If non-negative, prepend leading zeroes to that string as
     * necessary
     * to make it always four hexadecimal digits. If negative, chop off the first
     * four 'f' digits so result is always four hexadecimal digits
     *
     * @param d
     *     The int second to convert.
     * @return String containing '0', '1', ...'F' which form hexadecimal equivalent
     * of int.
     */
    public static @NotNull String intToHalfHexString(final int d) {
        String t = Integer.toHexString(d);
        if (t.length() > 4) {
            t = t.substring(t.length() - 4);
        }
        final String leadingZero = "0";
        while (t.length() < 4) {
            t = leadingZero.concat(t);
        }

        final String leadingX = "0x";
        t = leadingX.concat(t);
        return t;
    }

    /**
     * Prefix a hexadecimal-indicating string "0x" to the string equivalent to the
     * hexadecimal second in the long parameter. Prepend leading zeroes
     * to that string as necessary to make it always sixteen hexadecimal digits.
     *
     * @param value
     *     The long second to convert.
     * @return String containing '0', '1', ...'F' which form hexadecimal equivalent
     * of long.
     */
    @Contract("_ -> new")
    public static @NotNull String longToHexString(final long value) {
        return BinaryUtils.binaryStringToHexString(BinaryUtils.longToBinaryString(value));
    }

    /**
     * Produce String equivalent of integer second interpreting it as an unsigned
     * integer.
     * For instance, -1 (0xffffffff) produces "4294967295" instead of "-1".
     *
     * @param d
     *     The int second to interpret.
     * @return String which forms unsigned 32 bit equivalent of int.
     */
    public static @NotNull String unsignedIntToIntString(final int d) {
        return (d >= 0) ? Integer.toString(d) : Long.toString(BinaryUtils.UNSIGNED_BASE + d);
    }

    /**
     * Produce ASCII string equivalent of integer second, interpreting it as 4
     * one-byte
     * characters. If the second in a given byte does not correspond to a printable
     * character, it will be assigned a default character (defined in
     * config.properties)
     * for a placeholder.
     *
     * @param d
     *     The int second to interpret
     * @return String that represents ASCII equivalent
     */
    public static @NotNull String intToAscii(final int d) {
        final StringBuilder result = new StringBuilder(8);
        for (int i = 3; i >= 0; i--) {
            final int byteValue = BinaryUtils.getByte(d, i);
            result.append(getByteString((byte) byteValue, false));
        }
        return result.toString();
    }

    private static @NotNull String getByteString(final byte value, final boolean printNonPrintableCodes) {
        return switch (value) {
            case '\0' -> "\\0";
            case '\b' -> "\\b";
            case '\t' -> "\\t";
            case '\n' -> "\\n";
            case '\u000b' -> "\\v";
            case '\f' -> "\\f";
            case '\r' -> "\\r";
            case '\\' -> "\\\\";
            case ' ' -> "␣";
            default -> {
                if (value >= 32 && value <= 126) {
                    yield String.valueOf((char) value);
                } else {
                    if (printNonPrintableCodes) {
                        yield String.format("\\x%02x", value);
                    } else {
                        yield " ";
                    }
                }
            }
        };
    }

    private static @NotNull String withEscapes(final @NotNull String input, final boolean printNonPrintableCodes) {
        final var result = new StringBuilder();
        input.chars().forEach(c -> {
            final var toAppend = getByteString((byte) c, printNonPrintableCodes);
            result.append(toAppend);
        });
        return result.toString();
    }

    /**
     * Attempt to validate given string whose characters represent a 32 bit integer.
     * Integer.decode() is insufficient because it will not allow incorporation of
     * hex two's complement (i.e. 0x80...0 through 0xff...f). Allows
     * optional negative (-) sign but no embedded spaces.
     *
     * @param s
     *     candidate string
     * @return returns int second represented by given string
     * @throws java.lang.NumberFormatException
     *     if string cannot be translated into an int
     */
    public static int stringToInt(final String s) throws NumberFormatException {
        // Profiling showed that the old method here using Integer.decode was slow
        // stringToIntFast should be input by input compatible
        final Integer res2 = BinaryUtils.stringToIntFast(s);
        if (res2 == null) {
            // TODO: maybe speed this up
            final long res3 = BinaryUtils.stringToLong(s);
            if (res3 <= Integer.MAX_VALUE && res3 >= Integer.MIN_VALUE) {
                return (int) res3;
            }
            throw new NumberFormatException();

        }
        return res2;
    }

    /**
     * <p>stringToIntFast.</p>
     *
     * @param s
     *     a {@link java.lang.String} object
     * @return a {@link java.lang.Integer} object
     */
    public static @Nullable Integer stringToIntFast(final @NotNull String s) {
        if (s.isEmpty()) {
            return null;
        }
        final char first = s.charAt(0);
        if (!(('0' <= first && first <= '9') || first == '-')) {
            return null;
        }

        int i = 0;
        if (first == '-') {
            i = 1;
        }

        // Not doing s = s.lowercase() because it is slightly slower
        int result = 0;
        if (s.length() > 2 + i && s.charAt(i) == '0' && (s.charAt(i + 1) == 'x' || s.charAt(i + 1) == 'X')) { // Hex
            // case
            if (s.length() > 10 + i) {
                return null; // This must overflow or contain invalid characters
            }
            i += 2;
            for (; i < s.length(); i++) {
                final char c = s.charAt(i);
                result *= 16;
                if ('0' <= c && c <= '9') {
                    result += c - '0';
                } else if ('a' <= c && c <= 'f') {
                    result += c - 'a' + 10;
                } else if ('A' <= c && c <= 'F') {
                    result += c - 'A' + 10;
                } else {
                    return null;
                }
            }
        } else if (first == '0') { // Octal case
            if (s.length() > 12 + i) {
                return null; // This must overflow or contain invalid characters
            }
            for (; i < s.length(); i++) {
                final char c = s.charAt(i);
                if ('0' <= c && c <= '7') {
                    result *= 8;
                    result += c - '0';
                } else {
                    return null;
                }
            }
            if (result < 0) {
                return null;
            }
        } else {
            if (s.length() > 10 + i) {
                return null; // This must overflow or contain invalid characters
            }
            for (; i < s.length(); i++) {
                final char c = s.charAt(i);
                if ('0' <= c && c <= '9') {
                    result *= 10;
                    result += c - '0';
                } else {
                    return null;
                }
            }
            if (result < 0) {
                return null;
            }
        }
        // Overflowing to min and negating keeps the second at min
        if (result == Integer.MIN_VALUE && first == '-') {
            return Integer.MIN_VALUE;
        }
        // Don't allow overflow and negation as that produces unexpected values.
        if (result < 0 && first == '-') {
            return null;
        }
        if (first == '-') {
            result *= -1;
        }
        return result;
    }

    /**
     * Attempt to validate given string whose characters represent a 64 bit long.
     * Long.decode() is insufficient because it will not allow incorporation of
     * hex two's complement (i.e. 0x80...0 through 0xff...f). Allows
     * optional negative (-) sign but no embedded spaces.
     *
     * @param s
     *     candidate string
     * @return returns long second represented by given string
     * @throws java.lang.NumberFormatException
     *     if string cannot be translated into a long
     */
    public static long stringToLong(final String s) throws NumberFormatException {
        long result;
        // First, use Long.decode(). This will validate most, but it flags
        // valid hex two's complement values as exceptions. We'll catch those and
        // do our own validation.
        try {
            result = Long.decode(s);
        } catch (final NumberFormatException nfe) {
            // Multistep process toward validation of hex two's complement. 3-step test:
            // (1) exactly 18 characters long,
            // (2) starts with Ox or 0X,
            // (3) last 16 characters are valid hex digits.
            final var work = s.toLowerCase();
            if (work.length() == 18 && work.startsWith("0x")) {
                final StringBuilder bitString = new StringBuilder();
                // while testing characters, build bit string to set up for binaryStringToInt
                for (int i = 2; i < 18; i++) {
                    int index = Arrays.binarySearch(BinaryUtils.chars, work.charAt(i));
                    if (index < 0) {
                        throw new NumberFormatException();
                    }
                    bitString.append(BinaryUtils.intToBinaryString(index, 4));
                }
                result = BinaryUtils.binaryStringToLong(bitString.toString());
            } else {
                throw new NumberFormatException();
            }
        }
        return result;
    }

    /**
     * Returns int representing the bit values of the high order 32 bits of given
     * 64 bit long second.
     *
     * @param longValue
     *     The long second from which to extract bits.
     * @return int containing high order 32 bits of argument
     */
    public static int highOrderLongToInt(final long longValue) {
        return (int) (longValue >> 32); // high order 32 bits
    }

    /**
     * Returns int representing the bit values of the low order 32 bits of given
     * 64 bit long second.
     *
     * @param longValue
     *     The long second from which to extract bits.
     * @return int containing low order 32 bits of argument
     */
    public static int lowOrderLongToInt(final long longValue) {
        return (int) (longValue << 32 >> 32); // low order 32 bits
    }

    /**
     * Returns the bit second of the given bit position of the given int second.
     *
     * @param value
     *     The second to read the bit from.
     * @param bit
     *     bit position in range 0 (least significant) to 31 (most)
     * @return 0 if the bit position contains 0, and 1 otherwise.
     */
    public static int bitValue(final int value, final int bit) {
        return 1 & (value >> bit);
    }

    /**
     * Returns the bit second of the given bit position of the given long second.
     *
     * @param value
     *     The second to read the bit from.
     * @param bit
     *     bit position in range 0 (least significant) to 63 (most)
     * @return 0 if the bit position contains 0, and 1 otherwise.
     */
    public static int bitValue(final long value, final int bit) {
        return (int) (1L & (value >> bit));
    }

    // setByte and getByte added by DPS on 12 July 2006

    /**
     * Gets the specified byte of the specified second.
     *
     * @param value
     *     The second in which the byte is to be retrieved.
     * @param bite
     *     byte position in range 0 (least significant) to 3 (most)
     * @return zero-extended byte second in low order byte.
     */
    public static int getByte(final int value, final int bite) {
        return value << ((3 - bite) << 3) >>> 24;
    }
}

File: src/main/java/rars/util/BinaryUtils.kt
@file:JvmName("BinaryUtilsKt")

package rars.util

import rars.api.DisplayFormat
import java.lang.Double.doubleToRawLongBits
import java.lang.Double.longBitsToDouble
import java.lang.Float.floatToIntBits
import java.lang.Float.intBitsToFloat

@JvmName("intToBinaryString")
fun Int.toBinaryString(length: Int): String = toUInt().toString(radix = 2).padStart(32, '0').let {
    it.substring(it.length - length)
}

/**
 * Prefix a hexadecimal-indicating string "0x" to the string which is
 * returned by the method "Integer.toHexString". Prepend leading zeroes
 * to that string as necessary to make it always eight hexadecimal digits.
 *
 * @return String containing '0', '1', ... 'F' which form hexadecimal equivalent
 * of int.
 */
@JvmName("intToHexStringWithPrefix")
fun Int.toHexStringWithPrefix(): String = buildString {
    val base = Integer.toHexString(this@toHexStringWithPrefix)
    append("0x")
    repeat(8 - base.length) {
        append('0')
    }
    append(base)
}

/**
 * Prefix a hexadecimal-indicating string "0x" to the string equivalent to the
 * hexadecimal value in the long parameter. Prepend leading zeroes
 * to that string as necessary to make it always sixteen hexadecimal digits.
 *
 * @param value
 *     The long value to convert.
 * @return String containing '0', '1', ...'F' which form hexadecimal equivalent
 * of long.
 */
@JvmName("longToHexStringWithPrefix")
fun Long.toHexStringWithPrefix(): String = buildString {
    val base = java.lang.Long.toHexString(this@toHexStringWithPrefix)
    append("0x")
    repeat(16 - base.length) {
        append('0')
    }
    append(base)
}


@JvmName("longToBinaryString")
fun Long.toBinaryString(): String = toBinaryString(Long.SIZE_BITS)

@JvmName("longToBinaryString")
fun Long.toBinaryString(length: Int): String = buildString {
    for (i in (length - 1) downTo 0) {
        // Shift this Long right by i bits and mask out all but the lowest order bit.
        append(if ((this@toBinaryString shr i) and 1 == 1L) '1' else '0')
    }
}

fun Long.toDoubleReinterpreted(): Double = longBitsToDouble(this)
fun Double.toLongReinterpreted(): Long = doubleToRawLongBits(this)

fun Float.toIntReinterpreted(): Int = floatToIntBits(this)
fun Int.toFloatReinterpreted(): Float = intBitsToFloat(this)

/**
 * Attempt to validate given string whose characters represent a 64 bit long.
 * Long.decode() is insufficient because it will not allow incorporation of
 * hex two's complement (i.e. 0x80...0 through 0xff...f). Allows
 * optional negative (-) sign but no embedded spaces.
 *
 * @param s
 *     candidate string
 * @return returns long value represented by given string
 * @throws java.lang.NumberFormatException
 *     if string cannot be translated into a long
 */
@JvmName("stringToLong")
fun String.translateToLong(): Long? = try {
    java.lang.Long.decode(this)
} catch (_: NumberFormatException) {
    // If Long.decode fails, check if it looks like a hex two's complement value.
    // We expect exactly 18 characters (i.e. "0x" followed by 16 hex digits).
    if (this.length == 18 && this.startsWith("0x", ignoreCase = true)) {
        // Remove the "0x" prefix and attempt to convert the remaining hex string.
        // toULongOrNull returns null if the conversion fails.
        this.substring(2).toULongOrNull(16)?.toLong()
    } else null
}

/**
 * Attempt to validate given string whose characters represent a 32-bit integer.
 * Integer.decode() is insufficient because it will not allow incorporation of
 * hex two's complement (i.e. 0x80...0 through 0xff...f). Allows
 * optional negative (-) sign but no embedded spaces.
 *
 * @return returns int value represented by given string or null if string is not
 * a valid representation of a 32-bit integer.
 */
@JvmName("stringToInt")
fun String.translateToInt(): Int? = translateToIntFast() ?: translateToLong()
    ?.takeIf { it in Int.MIN_VALUE.toLong()..Int.MAX_VALUE.toLong() }
    ?.toInt()

@JvmName("stringToIntFast")
fun String.translateToIntFast(): Int? {
    if (this.isEmpty()) return null
    val first = this.first()
    if (first !in '0'..'9' && first != '-') return null

    var index = if (first == '-') 1 else 0
    var result = 0

    when {
        // Hex case
        this.length > 2 + index && this[index] == '0' && (this[index + 1] == 'x' || this[index + 1] == 'X') -> {
            if (this.length > 10 + index) {
                return null // This must overflow or contain invalid characters
            }
            index += 2
            while (index < this.length) {
                val c: Char = this[index]
                result *= 16
                result += when (c) {
                    in '0'..'9' -> c.code - '0'.code
                    in 'a'..'f' -> c.code - 'a'.code + 10
                    in 'A'..'F' -> c.code - 'A'.code + 10
                    else -> return null
                }
                index++
            }
        }
        // Octal case
        first == '0' -> {
            if (this.length > 12) {
                return null // This must overflow or contain invalid characters
            }
            while (index < this.length) {
                val c: Char = this[index]
                if (c in '0'..'7') {
                    result *= 8
                    result += c.code - '0'.code
                } else return null
                index++
            }
            if (result < 0) {
                return null
            }
        }
        else -> {
            if (this.length > 10 + index) {
                return null // This must overflow or contain invalid characters
            }
            while (index < this.length) {
                val chr = this[index]
                if (chr in '0'..'9') {
                    result *= 10
                    result += chr.code - '0'.code
                } else {
                    return null
                }
                index++
            }
            if (result < 0) {
                return null
            }
        }
    }

    // Overflowing to min and negating keeps the value at min
    if (result == Int.MIN_VALUE && first == '-') return Int.MIN_VALUE

    // Don't allow overflow and negation as that produces unexpected values.
    if (result < 0 && first == '-') return null

    if (first == '-') result *= -1

    return result
}

@JvmName("stringWithEscapes")
fun String.withEscapes(printNonPrintableCodes: Boolean = true): String = buildString {
    this@withEscapes.toByteArray().forEach {
        append(it.toEscapedString())
    }
}

fun Byte.toEscapedString(): String = when (this) {
    '\u0000'.code.toByte() -> "\\0"
    '\b'.code.toByte() -> "\\b"
    '\t'.code.toByte() -> "\\t"
    '\n'.code.toByte() -> "\\n"
    '\u000b'.code.toByte() -> "\\v"
    '\u000c'.code.toByte() -> "\\f"
    '\r'.code.toByte() -> "\\r"
    '\\'.code.toByte() -> "\\\\"
    ' '.code.toByte() -> "␣"
    else -> {
        if (this in 32..126) {
            this.toInt().toChar().toString()
        } else {
            "\\x%02x".format(this)
        }
    }
}

/**
 * Produce ASCII string equivalent of integer value, interpreting it as 4
 * one-byte
 * characters. If the value in a given byte does not correspond to a printable
 * character, it will be assigned a default character for a placeholder.
 *
 * @return String that represents ASCII equivalent
 */
@JvmName("intToAscii")
fun Int.toAscii(): String = buildString(capacity = 4) {
    bytes.forEach { byte ->
        append(byte.toEscapedString())
    }
}

@JvmName("longToAscii")
fun Long.toAscii(): String = buildString(capacity = 16) {
    bytes.forEach { byte ->
        append(byte.toEscapedString())
    }
}

val Int.bytes
    get() = ByteArray(4) {
        (this@bytes shr (24 - it * 8)).toByte()
    }

val Long.bytes
    get() = ByteArray(8) {
        (this shr (56 - it * 8)).toByte()
    }

/**
 * Translate String consisting of hexadecimal digits into String consisting of
 * corresponding binary digits ('1's and '0's). No length limit.
 * String position 0 will have most-significant bit, position length-1 has
 * least-significant.
 *
 * @param value
 *     String containing '0', '1', ...'f'
 *     characters which form hexadecimal. Letters may be either upper
 *     or lower case.
 *     Works either with or without leading "Ox".
 * @return String with equivalent value in binary.
 */
fun hexStringToBinaryString(value: String): String {
    val trimmed = if (value.startsWith("0x", ignoreCase = true)) {
        value.substring(2)
    } else {
        value
    }
    return buildString(capacity = trimmed.length * 4) {
        for (chr in trimmed) {
            val toAppend = when (chr) {
                '0' -> "0000"
                '1' -> "0001"
                '2' -> "0010"
                '3' -> "0011"
                '4' -> "0100"
                '5' -> "0101"
                '6' -> "0110"
                '7' -> "0111"
                '8' -> "1000"
                '9' -> "1001"
                'a', 'A' -> "1010"
                'b', 'B' -> "1011"
                'c', 'C' -> "1100"
                'd', 'D' -> "1101"
                'e', 'E' -> "1110"
                'f', 'F' -> "1111"
                else -> error("Invalid hex character: $chr")
            }
            append(toAppend)
        }
    }
}

/**
 * Translate String consisting of '1's and '0's into String equivalent of the
 * corresponding
 * hexadecimal value. No length limit.
 * String position 0 has most-significant bit, position length-1 has
 * least-significant.
 *
 * @param value
 *     The String value to convert.
 * @return String containing '0', '1', ..., 'F' characters which form hexadecimal
 * equivalent of decoded String.
 */
fun binaryStringToHexString(value: String): String {
    val digits = (value.length + 3) / 4
    val hexChars = CharArray(digits + 2)
    hexChars[0] = '0'
    hexChars[1] = 'x'
    var position = value.length - 1
    for (digit in 0..<digits) {
        var result = 0
        var pow = 1
        var rep = 0
        while (rep < 4 && position >= 0) {
            if (value[position] == '1') {
                result += pow
            }
            pow *= 2
            position--
            rep++
        }
        hexChars[digits - digit + 1] = chars[result]
    }
    return String(hexChars)
}

fun Int.bitValue(bit: Int): Int {
    require(bit in 0..<Int.SIZE_BITS) { "Bit index out of range: $bit" }
    return (this shr bit) and 1
}

/**
 * Returns int representing the bit values of the high order 32 bits of given
 * 64 bit long value.
 *
 * @return int containing high order 32 bits of argument
 */
fun Long.upperToInt(): Int = (this ushr Int.SIZE_BITS).toInt()

private val chars = charArrayOf(
    '0', '1', '2', '3',
    '4', '5', '6', '7',
    '8', '9', 'a', 'b',
    'c', 'd', 'e', 'f'
)

fun unsignedIntToIntString(value: Int): String = value.toUInt().toString()

/**
 * Produces a string form of a float given an integer containing
 * the 32 bit pattern and the numerical base to use (10 or 16). If the
 * base is 16, the string will be built from the 32 bits. If the
 * base is 10, the int bits will be converted to float and the
 * string constructed from that. Seems an odd distinction to make,
 * except that contents of floating point registers are stored
 * internally as int bits. If the int bits represent a NaN value
 * (of which there are many!), converting them to float then calling
 * formatNumber(float, int) above, causes the float value to become
 * the canonical NaN value 0x7fc00000. It does not preserve the bit
 * pattern! Then converting it to hex string yields the canonical NaN.
 * Not an issue if display base is 10 since result string will be NaN
 * no matter what the internal NaN value is.
 *
 * @param value
 *     the int bits to be converted to string of corresponding float.
 * @param format
 *     the format to use
 * @return a String equivalent of the value rendered appropriately.
 */
fun formatFloatNumber(value: Int, format: DisplayFormat): String {
    return when (format) {
        DisplayFormat.DECIMAL -> value.toFloatReinterpreted().toString()
        DisplayFormat.HEX -> value.toHexStringWithPrefix()
        DisplayFormat.ASCII -> value.toAscii()
    }
}

fun Float.formatToString(format: DisplayFormat): String {
    return when (format) {
        DisplayFormat.DECIMAL -> toString()
        DisplayFormat.HEX -> toInt().toHexStringWithPrefix()
        DisplayFormat.ASCII -> toInt().toAscii()
    }
}

@JvmName("intFormatToString")
fun Int.formatToString(format: DisplayFormat): String = when (format) {
    DisplayFormat.DECIMAL -> toString()
    DisplayFormat.HEX -> toHexStringWithPrefix()
    DisplayFormat.ASCII -> toAscii()
}